File: 2016/day19/part2.py
# NUM_ELVES = 5
NUM_ELVES = 3018458

class Elve:
    def __init__(self, i, p, n):
        self.i = i
        self.prev = p
        self.next = n

head = e = Elve(1, None, None)
target = None
while e.i < NUM_ELVES:
    e.next = Elve(e.i+1, e, None)
    e = e.next
    if e.i == (NUM_ELVES // 2) + 1:
        target = e
head.prev = e
e.next = head

size = NUM_ELVES
while head.next != head:
    target.prev.next = target.next
    target.next.prev = target.prev
    if size % 2 == 0:
        target = target.next
    else:
        target = target.next.next
    head = head.next
    size -= 1

print(head.i)

File: 2018/day9/part2.clj
(def nplayers 478)
(def last-marble (* 100 71240))

(defn create-circle
  []
  (let [head (transient {:value 0})]
    (assoc! head :next head)
    (assoc! head :prev head)
    head))

(defn move
  [current steps direction]
  (loop [node current 
         step 0]
    (if (= step steps)
      node
      (recur (direction node) (inc step)))))

(defn insert
  [current value]
  (let [next1 (move current 1 :next)
        next2 (move current 2 :next)
        new-node (transient {:value value})]
    (assoc! next1 :next new-node)
    (assoc! next2 :prev new-node)
    (assoc! new-node :prev next1)
    (assoc! new-node :next next2)
    new-node))

(defn remove
  [current]
  (let [prev (:prev current)
        next (:next current)]
    (assoc! prev :next next)
    (assoc! next :prev prev)
    next))

(defn max-score
  []
  (loop [circle (create-circle) 
         current-player -1
         scores (vec (repeat nplayers 0))
         marbles (drop 1 (range))]
    (let [marble (first marbles)]
      (if (zero? (mod marble 23))
        (let [to-remove (move circle 7 :prev)]
          (recur (remove to-remove)
                 (mod (inc current-player) nplayers)
                 (update scores
                         current-player #(+ % marble (:value to-remove)))
                 (rest marbles)))
        (if (<= marble last-marble)
          (recur (insert circle marble)
                 (mod (inc current-player) nplayers)
                 scores
                 (rest marbles))
          (apply max scores))))))